File: edit_distance.py
class Solution:
    def helper(self, w1, w2, w1_curr, w2_curr, d):
        if (w1_curr, w2_curr) in d:
            return d[(w1_curr, w2_curr)]
        if(w1_curr >= len(w1)):
            return len(w2) - w2_curr
        elif(w2_curr >= len(w2)):
            return len(w1) - w1_curr
        
        if w1[w1_curr] == w2[w2_curr]:
            d[(w1_curr, w2_curr)] = self.helper(w1,w2, w1_curr+1, w2_curr+1, d)
        else:
            d[(w1_curr, w2_curr)] = min(self.helper(w1, w2, w1_curr+1, w2_curr, d),
                                        self.helper(w1, w2, w1_curr+1, w2_curr+1, d), 
                                        self.helper(w1, w2, w1_curr, w2_curr+1, d)) + 1
        return d[(w1_curr, w2_curr)]
        
    def minDistance(self, word1: str, word2: str) -> int:
        d = {}
        return self.helper(word1, word2, 0, 0, d)

File: edit_distance.cpp
#include <string>
#include <vector>
using namespace std;

int minDistance(string word1, string word2)
{
    int n1 = word1.size();
    int n2 = word2.size();
    if(!n1 || !n2)
        return max(n1, n2);
    vector<vector<int>> dp(n1, vector<int>(n2));
    
    for(int i = 0; i < n1; i++)
    {
        for(int j = 0; j < n2; j++)
        {
            if(word1[i] == word2[j])
                dp[i][j] = i && j ? dp[i-1][j-1] : max(i, j);
            else
            {
                int insert = j ? dp[i][j-1] : i;
                int del = i ? dp[i-1][j] : j;
                int replace = i && j ? dp[i-1][j-1] : max(i, j);
                dp[i][j] = min({insert, del, replace}) + 1;
            }
        }
    }
    return dp[n1-1][n2-1];
}

int main()
{

}